src/RestrictedTokenFaucet.sol

```diff
@@ -0,0 +1,59 @@
+pragma solidity >=0.5.0;
+
+interface ERC20Like {
+    function balanceOf(address) external view returns (uint256);
+    function transfer(address,uint256) external; // return bool?
+}
+
+contract RestrictedTokenFaucet {
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address guy) public isOwner { wards[guy] = 1; }
+    function deny(address guy) public isOwner { wards[guy] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+    modifier isOwner { require(owner == msg.sender); _; }
+    address public owner;
+
+    uint256 public amt;
+    mapping (address => mapping (address => bool)) public done;
+
+    constructor (uint256 amt_) public {
+        wards[msg.sender] = 1;
+        owner = msg.sender;
+        amt = amt_;
+    }
+
+    function mul(uint x, uint y) internal pure returns (uint z) {
+        require(y == 0 || (z = x * y) / y == x);
+    }
+
+    function gulp(address gem) external auth {
+        require(!done[msg.sender][address(gem)], "token-faucet: already used faucet");
+        require(ERC20Like(gem).balanceOf(address(this)) >= amt, "token-faucet: not enough balance");
```

Comment:
Why not check the return value of `transfer` instead?

Reply:
This was from the regular `token-faucet` but my guess would be because ERC-20s cannot always be expected to return according to the spec.  Since this was designed as a multi-token faucet, I'm guessing this provides flexibility without the need for adapters.